File: 06 Approximation for Model Inference/MCMC.R
# same as Risk_Measure.R
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))

################################################################################
library(mvtnorm)

sigmoid <- function(beta, X, y=NA){
  exp_eta <- exp(X %*% beta)
  return(exp_eta / (1 + exp_eta))
}

MCMC_logistic <- function(X, y, num_it=1e5, nu=0.5, seed=4002){
  # ln P(beta): large cov to be a non-informative prior
  log_prior <- function(beta) sum(dnorm(beta, 0, 100, log=TRUE))
  
  log_likelihood <- function(beta, X, y){   # ln L(beta)
    pi <- sigmoid(beta, X, y)
    return(sum(y * log(pi) + (1 - y) * log(1 - pi)))
  }
  
  log_target <- function(beta, X, y){       # approx. ln P(beta| X, y)
    return(log_likelihood(beta, X, y) + log_prior(beta))
  }
  
  set.seed(seed)
  X_ <- cbind(1, as.matrix(X))
  n_accept <- 0
  
  Cov_nu <- nu * solve(t(X_) %*% X_)
  beta_hist <- matrix(NA, nrow=num_it, ncol=ncol(X_))
  beta_hist[1,] <- rmvnorm(1, rep(0, ncol(X_)), Cov_nu)
  for (i in 2:num_it){
    beta_old <- beta_hist[i-1,]
    beta_new <- rmvnorm(1, beta_old, Cov_nu)
    
    log_target_old <- log_target(matrix(beta_old, ncol=1), X_, y)
    log_target_new <- log_target(t(beta_new), X_, y)
    
    alpha <- exp(log_target_new - log_target_old)
    if (!is.na(alpha)){
      if (runif(1) < min(1, alpha)){
        beta_old <- beta_new
        n_accept <- n_accept + 1
      }
    }
    
    beta_hist[i,] <- beta_old
  }
  return(list(beta_hist=beta_hist, n_accept=n_accept, num_it=num_it))
}

get_coefficient <- function(MCMC_lst, method="median", burn_in=0){
  beta_hist <- MCMC_lst$beta_hist
  func <- get(method)
  return(apply(beta_hist[-c(1:burn_in),], MARGIN=2, FUN=func))
}

MCMC_predict <- function(X, MCMC_lst, method="median", burn_in=0){
  X_ <- cbind(1, as.matrix(X))
  beta <- get_coefficient(MCMC_lst, method, burn_in)
  pi <- sigmoid(beta, X_)
  return(as.numeric(pi > 0.5))
}

################################################################################
df <- read.csv("fin-ratio.csv")   # read in data file
X <- subset(df, select=-c(HSI)); y <- df$HSI

MCMC_lst <- MCMC_logistic(X, y, num_it=5e5)
burn_in <- 1e5
y_hat_MCMC <- MCMC_predict(X, MCMC_lst, method="median", 
                           burn_in=burn_in)
MCMC_lst$n_accept / MCMC_lst$num_it
cat("MCMC_median", get_coefficient(MCMC_lst, "median", 
                                   burn_in=burn_in), "\n")
cat("MCMC_median", get_coefficient(MCMC_lst, "mean", 
                                   burn_in=burn_in), "\n")

################################################################################
glm_model <- glm(HSI ~ ., data=df, family="binomial")
y_hat_MLE <- as.numeric(glm_model$fitted.values > 0.5)
cat("MLE", coef(glm_model), "\n")

################################################################################
which(y == 1)
cat("MCMC_median:", mean(y == y_hat_MCMC), "\n")
which(y_hat_MCMC == 1)
cat("MLE:", mean(y == y_hat_MLE), "\n")
which(y_hat_MLE == 1)

################################################################################
MCMC_median <- get_coefficient(MCMC_lst, "median", burn_in=burn_in)
beta_median <- MCMC_lst$beta_hist
par(mfrow=c(1, 2))
for (col in 1:(ncol(X)+1)){
  plot(beta_median[,col], type="l", lwd=2, main="MCMC path", 
       xlab="iterations", ylab=paste("beta", col-1))
  abline(v=burn_in, lty=2, lwd=2, col="blue")
  abline(h=MCMC_median[col], lty=2, lwd=2, col="red")
  
  hist(beta_median[-c(1:burn_in), col], main="MCMC histogram",
       xlab=paste("beta", col-1))
  abline(v=MCMC_median[col], lty=2, lwd=2, col="red")
}

File: 06 Approximation for Model Inference/MCMC.py
import numpy as np
from scipy.stats import multivariate_normal
from tqdm import trange
import matplotlib.pyplot as plt

#%%
class MCMC_logistic:
    def __init__(self, num_it=5e4, nu=0.5, seed=4002):
        self.num_it = int(num_it)
        self.nu = nu
        self.seed = seed
        self.proposal_rvs = multivariate_normal.rvs
        
        self.burn_in = None
    
    def sigmoid(self, beta, X, y=None):
        exp_eta = np.exp(X @ beta)
        return (exp_eta / (1 + exp_eta)).reshape(y.shape)

    def log_prior(self, beta):              # ln P(beta)
        # large cov to be a non-informative prior
        return np.sum(multivariate_normal.logpdf(beta, mean=0, 
                                                 cov=100**2))
    
    def log_likelihood(self, beta, X, y):   # ln L(beta) 
        pi = self.sigmoid(beta, X, y)
        return np.sum(y * np.log(pi) + (1 - y) * np.log(1 - pi))
    
    def log_target(self, beta, X, y):       # approx. ln P(beta| X, y)
        return self.log_likelihood(beta, X, y) + self.log_prior(beta)
    
    def fit(self, X, y):
        np.random.seed(self.seed)
        X_ = np.c_[np.ones(len(y)), X]       # add intercept
        self.n_accept = 0
        self.n_col = X_.shape[1]
        
        Cov_nu = self.nu * np.linalg.inv(X_.T @ X_)
        beta_hist = np.empty(shape=(self.num_it, self.n_col))
        beta_hist[1] = self.proposal_rvs(mean=np.repeat(0,self.n_col), 
                                         cov=Cov_nu, size=1)
        for i in trange(1, self.num_it):
            beta_old = beta_hist[i-1]
            beta_new = self.proposal_rvs(mean=beta_old, 
                                         cov=Cov_nu, 
                                         size=1).reshape(-1, 1)
            
            # Since the proposal distribution, normal, is symmetric, 
            # we have g(beta_old| beta_new) = g(beta_new| beta_old)
            log_target_old = self.log_target(beta_old, X_, y)
            log_target_new = self.log_target(beta_new, X_, y)
            
            u = np.random.rand()
            alpha = np.exp(log_target_new - log_target_old)
            # min(1, np.nan) = 1
            if u < min(1, alpha) and not np.isnan(alpha):
                beta_old = beta_new
                self.n_accept += 1
            
            beta_hist[i] = beta_old.T
        
        self.beta_dist = beta_hist
    
    def get_coefficient(self, method="median", burn_in=0):
        assert method in ["mean", "median"]
        func = getattr(np, method)
        return func(self.beta_dist[burn_in:], axis=0)
    
    def predict(self, X, method="median", burn_in=1e4):
        X_ = np.c_[np.ones(len(y)), X]
        self.burn_in = int(burn_in)
        beta = self.get_coefficient(method, self.burn_in)
        pi = self.sigmoid(beta, X_, np.arange(len(X_)))
        return (pi > 0.5).astype(np.int32)
    
    def plot(self, save_name, method="median", burn_in=None):
        if burn_in is None and self.burn_in is not None:
            burn_in = self.burn_in
        elif burn_in is None:
            burn_in = 0
        else:
            burn_in = int(burn_in)
            
        beta_MCMC = self.get_coefficient(method, int(burn_in))
        for col in range(self.n_col):
            fig, (ax1, ax2) = plt.subplots(ncols=2, 
                                           figsize=(20, 10))
            ax1.plot(self.beta_dist[:,col], color="black", 
                     linewidth=2)
            ax1.axvline(self.burn_in, color="b", linestyle="--", 
                        linewidth=2)
            ax1.axhline(beta_MCMC[col], color="r", linestyle="--", 
                        linewidth=2)
            ax1.set_xlabel("iterations", fontsize=15)
            ax1.set_ylabel(f"beta {col}", fontsize=15)
            ax1.set_title("MCMC path", fontsize=20)
            
            ax2.hist(self.beta_dist[burn_in:,col], alpha=0.5,
                     color="gray", ec='black', bins="sturges")
            ax2.axvline(beta_MCMC[col], color="r", linestyle="--")
            ax2.set_xlabel(f"beta {col}", fontsize=15)
            ax2.set_ylabel("frequencies", fontsize=15)
            ax2.set_title("MCMC histogram", fontsize=20)
            
            plt.tight_layout()
            fig.savefig(save_name + f"_beta {col}.png", dpi=200)
            

#%%
import pandas as pd

df = pd.read_csv("../Datasets/fin-ratio.csv")
X, y = df.drop(columns=["HSI"]).values, df.HSI.values

MCMC_model = MCMC_logistic(num_it=5e5)
MCMC_model.fit(X, y)
y_hat_MCMC = MCMC_model.predict(X, method="median", burn_in=1e5)
print(MCMC_model.n_accept / MCMC_model.num_it)
print("MCMC_median", MCMC_model.get_coefficient("median"))
print("MCMC_mean", MCMC_model.get_coefficient("mean"))

#%%
from sklearn.linear_model import LogisticRegression
sklearn_model = LogisticRegression(penalty=None)
sklearn_model.fit(X, y)
y_hat_MLE = sklearn_model.predict(X)
print("MLE", np.append(sklearn_model.intercept_, 
                       sklearn_model.coef_[0]))

#%%
print(np.where(y == 1)[0])
print(f"MCMC_median: {np.mean(y == y_hat_MCMC)}")
print(np.where(y_hat_MCMC == 1)[0])
print(f"MLE: {np.mean(y == y_hat_MLE)}")
print(np.where(y_hat_MLE == 1)[0])

#%%
MCMC_model.plot("../Picture/MCMC", method="median", 
                burn_in=None)
